### PR TITLE
cryptpad: init at v2.23.0

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

This project is part of the https://nlnet.nl/NGI/ grant.

The project still needs a bit of work to be ready to be merged:

* it's using bower which is deprecated by upstream
* it's tested with node 6.x (EOL) in the Dockerfile
* it assumes that the source folder is writable, it uses the source folder to store it's state
* it assumes that the configuration file is stored in a fixed location in the source root
* it assumes that the `./server.js` entry point is executed from the source root

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/web-apps/cryptpad.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+let
+  cfg = config.services.cryptpad;
+in
+{
+  options.services.cryptpad = {
+    enable = mkEnableOption "the Cryptpad service";
+
+    package = mkOption {
+      default = pkgs.cryptpad;
+      defaultText = "pkgs.nginxStable";
```
Comment (Infinisil) [inline nixos/modules/services/web-apps/cryptpad.nix]: Copy-paste error, probably

File: nixos/modules/services/web-apps/cryptpad.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        Cryptpad package to use.
+      ";
+    };
+
+    configPath = mkOption {
+      type = types.path;
+      default = "${cfg.package}/lib/node_modules/cryptpad/config/config.example.js";
+      defaultText = "\${cfg.package}/lib/node_modules/cryptpad/config/config.example.js";
+      description = ''
+        Path to the JavaScript configuration file.
```
Comment (Infinisil) [inline nixos/modules/services/web-apps/cryptpad.nix]: Can you link to the upstream docs for this config file here?

File: nixos/modules/services/web-apps/cryptpad.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+let
+  cfg = config.services.cryptpad;
+in
+{
+  options.services.cryptpad = {
+    enable = mkEnableOption "the Cryptpad service";
+
+    package = mkOption {
+      default = pkgs.cryptpad;
```
Comment (Infinisil) [inline nixos/modules/services/web-apps/cryptpad.nix]: I personally feel like we shouldn't add `package` options when there's no need to, but [not everyone feels the same](https://github.com/NixOS/nixpkgs/issues/50476).

File: nixos/modules/services/web-apps/cryptpad.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      description = "Cryptpad Service";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "networking.target" ];
+      serviceConfig = {
+        DynamicUser = true;
+        Environment = [
+          "CRYPTPAD_CONFIG=${cfg.configPath}"
+          "HOME=%S/cryptpad"
+        ];
+        ExecStart = "${pkgs.cryptpad}/bin/cryptpad";
```
Comment (Infinisil) [inline nixos/modules/services/web-apps/cryptpad.nix]: In any case, if you keep the `package` option, you should probably use it here

File: nixos/modules/services/web-apps/cryptpad.nix [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    package = mkOption {
+      default = pkgs.cryptpad;
+      defaultText = "pkgs.nginxStable";
+      type = types.package;
+      description = "
+        Cryptpad package to use.
+      ";
+    };
+
+    configPath = mkOption {
```
Comment (Infinisil) [inline nixos/modules/services/web-apps/cryptpad.nix]: Usually such options are called `configFile`

**Review by infinisil — CHANGES_REQUESTED** [review]
Review for the NixOS module

File: nixos/modules/services/web-apps/cryptpad.nix (hunk 1 above)
Comment (zimbatm) [reply nixos/modules/services/web-apps/cryptpad.nix]: thanks, fixed

File: nixos/modules/services/web-apps/cryptpad.nix (hunk 3 above)
Comment (zimbatm) [reply nixos/modules/services/web-apps/cryptpad.nix]: Ok let me contribute to that other thread then.

File: nixos/modules/services/web-apps/cryptpad.nix (hunk 4 above)
Comment (zimbatm) [reply nixos/modules/services/web-apps/cryptpad.nix]: thanks, fixed

File: nixos/modules/services/web-apps/cryptpad.nix (hunk 5 above)
Comment (zimbatm) [reply nixos/modules/services/web-apps/cryptpad.nix]: ok, changed

File: nixos/modules/services/web-apps/cryptpad.nix (hunk 2 above)
Comment (zimbatm) [reply nixos/modules/services/web-apps/cryptpad.nix]: fixed

File: nixos/modules/services/web-apps/cryptpad.nix (hunk 3 above)
Comment (zimbatm) [reply nixos/modules/services/web-apps/cryptpad.nix]: Looks like I already commented: https://github.com/NixOS/nixpkgs/issues/50476#issuecomment-439534060

**Pull request merged**